File: tools/richos-service/companion-windows/src/richos-companion/LiveCapture.cs
using System;
using System.Threading;
using RichOSCompanionCore;

namespace RichOSCompanion
{
    /// <summary>
    /// The live capture loop: wire the WASAPI engine to the <see cref="SessionWriter"/>, run a 10 Hz
    /// mix pump and a 1 Hz health tick + watchdog, and finalize cleanly on Ctrl-C. This is the
    /// outside-the-browser watchdog made real (§6.3) — it holds its own timer and alarms on stalls the
    /// capture can't hide. The peer of the macOS companion's <c>LiveCapture.swift</c>.
    /// </summary>
    internal static class LiveCapture
    {
        public static int Run(string zone, long startedAt, uint? pid, string? processName, string? processHint, string? supersedes)
        {
            // session.json is written at START (§6.1) — a call that captures nothing leaves a loud
            // open/empty session on disk, never an absence. Provisional method/target reflect intent and
            // are corrected once the engine resolves process-vs-system loopback.
            bool wantProcess = pid.HasValue;
            string method = wantProcess ? SessionContract.MethodProcessLoopback : SessionContract.MethodSystemLoopback;
            string target = wantProcess ? "process:" + (processName ?? pid!.Value.ToString()) : "system";

            var p = Program.MakeParams(startedAt, 48_000, method, target, processHint, supersedes);
            var writer = new SessionWriter(zone, p, Program.NowMs);
            try
            {
                writer.Start();
            }
            catch (Exception e)
            {
                Notifier.Alarm($"could not create the session directory: {e.Message}");
                return 1;
            }
            string dir = writer.Dir;
            Notifier.Info($"session (open) written: {dir}");

            var capture = new WasapiCapture(writer, pid, processName);
            try
            {
                capture.Start();
                Notifier.Info("capturing system audio + mic. Press Ctrl-C to stop.");
            }
            catch (Exception e)
            {
                // The open session.json stays on disk as the anomaly; we exit loud, not silent.
                Notifier.Alarm(
                    $"capture failed to start: {e.Message}. If the microphone is denied, allow it in " +
                    $"Settings > Privacy & security > Microphone, then retry. The open session at {dir} " +
                    "is the on-disk record that this call was NOT captured.");
                return 1;
            }

            // Correct the open session.json to the actually-resolved capture path (process vs system).
            if (capture.CaptureMethod != p.Method || capture.CaptureTarget != p.CaptureTarget)
            {
                p.Method = capture.CaptureMethod;
                p.CaptureTarget = capture.CaptureTarget;
                writer.RewriteOpenSessionJson();
            }

            var stop = new ManualResetEventSlim(false);
            int lastBytes = 0;
            int stalledSeconds = 0;

            // 10 Hz mix pump: keep the WAV growing and both sources frame-aligned.
            using var pump = new Timer(_ => { try { writer.Pump(); } catch { /* transient */ } }, null, 100, 100);

            // 1 Hz health tick + flush + watchdog escalation.
            using var health = new Timer(_ =>
            {
                var (mp, mr, sp, sr) = capture.SnapshotLevels();
                try
                {
                    writer.WriteHealthTick(mp, mr, sp, sr, capture.TapRunning, capture.MicRunning);
                    writer.Flush(); // durable to disk each second (§6.5)
                }
                catch { /* transient */ }

                // Watchdog: byte growth is the ground-truth liveness signal (§6.2 "no growth > 15 s").
                int bytes = writer.PumpedBytesTotal;
                if (bytes == lastBytes) stalledSeconds += 1; else stalledSeconds = 0;
                lastBytes = bytes;
                if (stalledSeconds == 15)
                {
                    Notifier.Alarm("no audio has been written for 15s — the capture has stalled (loopback "
                        + $"stopped, the render device changed, or the mic was revoked). Session: {dir}");
                }
            }, null, 1000, 1000);

            // Clean finalize on Ctrl-C: drain, flush, rewrite session.json as closed -> triggers P1.
            Console.CancelKeyPress += (_, e) => { e.Cancel = true; stop.Set(); };
            stop.Wait();

            pump.Dispose();
            health.Dispose();
            capture.Stop();
            try
            {
                writer.Close();
                Notifier.Info($"session (closed) written: {dir} — the P1 watcher will now transcribe it.");
            }
            catch (Exception e)
            {
                Notifier.Alarm($"failed to finalize session: {e.Message}");
            }
            return 0;
        }
    }
}

File: tools/richos-service/companion-macos/Sources/richos-companion/LiveCapture.swift
import Foundation
import RichOSCompanionCore

/// The live capture loop: wire the Core Audio engine to the SessionWriter, run a 10 Hz mix pump and
/// a 1 Hz health tick + watchdog, and finalize cleanly on Ctrl-C. This is the outside-the-browser
/// watchdog made real (§6.3) — it holds its own timer and alarms on stalls the capture can't hide.
@available(macOS 14.4, *)
func runLiveCapture(zone: String, startedAt: Int64) {
    // session.json is written at START (§6.1) — a call that captures nothing leaves a loud
    // open/empty session on disk, never an absence.
    let params = makeParams(startedAt: startedAt, sampleRate: 48_000)
    let writer = SessionWriter(rootZone: zone, params: params, clock: nowMs)
    do { try writer.start() } catch {
        Notifier.alarm("could not create the session directory: \(error.localizedDescription)")
        exit(1)
    }
    let dir = (zone as NSString).appendingPathComponent(params.sessionId)
    Notifier.info("session (open) written: \(dir)")

    let capture = CoreAudioCapture(writer: writer)
    do {
        try capture.start()
        Notifier.info("capturing system audio + mic. Press Ctrl-C to stop.")
    } catch {
        // The open session.json stays on disk as the anomaly; we exit loud, not silent.
        Notifier.alarm("""
        capture failed to start: \(error.localizedDescription). \
        If this is the first run, grant "System Audio Recording" and Microphone in \
        System Settings > Privacy & Security, then retry. The open session at \(dir) is the on-disk \
        record that this call was NOT captured.
        """)
        exit(1)
    }

    let timerQueue = DispatchQueue(label: "com.richos.companion.timers")

    // 10 Hz mix pump: keep the WAV growing and both sources frame-aligned.
    let pump = DispatchSource.makeTimerSource(queue: timerQueue)
    pump.schedule(deadline: .now() + 0.1, repeating: 0.1)
    pump.setEventHandler { _ = try? writer.pump() }
    pump.resume()

    // 1 Hz health tick + fsync + watchdog escalation.
    var lastBytes = 0
    var stalledSeconds = 0
    let health = DispatchSource.makeTimerSource(queue: timerQueue)
    health.schedule(deadline: .now() + 1.0, repeating: 1.0)
    health.setEventHandler {
        let (mp, mr, sp, sr) = capture.snapshotLevels()
        try? writer.writeHealthTick(
            micPeak: mp, micRms: mr, sysPeak: sp, sysRms: sr,
            tapRunning: capture.tapRunning, micRunning: capture.micRunning)
        try? writer.flush() // durable to disk each second (§6.5)

        // Watchdog: byte growth is the ground-truth liveness signal (§6.2 "no growth > 15 s").
        let bytes = writer.pumpedBytesTotal
        if bytes == lastBytes { stalledSeconds += 1 } else { stalledSeconds = 0 }
        lastBytes = bytes
        if stalledSeconds == 15 {
            Notifier.alarm("no audio has been written for 15s — the capture has stalled (tap stopped, "
                + "device changed, or System Audio Recording permission was revoked). Session: \(dir)")
        }
    }
    health.resume()

    // Clean finalize on SIGINT: drain, flush, rewrite session.json as closed -> triggers P1.
    signal(SIGINT, SIG_IGN)
    let sigint = DispatchSource.makeSignalSource(signal: SIGINT, queue: timerQueue)
    sigint.setEventHandler {
        pump.cancel(); health.cancel()
        capture.stop()
        do {
            try writer.close()
            Notifier.info("session (closed) written: \(dir) — the P1 watcher will now transcribe it.")
        } catch {
            Notifier.alarm("failed to finalize session: \(error.localizedDescription)")
        }
        exit(0)
    }
    sigint.resume()

    RunLoop.main.run()
}
